### PR TITLE
route_schedule: Improve ranked pairs implementation

### PR DESCRIPTION
**Pull request by TeXitoi**
http://jira.canaltp.fr/browse/NAVITIAII-1988

Bench for RER and Transilien on fr-idf (kraken only):

| Line | removed | topo before | time before | topo after | time after |
| --- | --- | --- | --- | --- | --- |
| A | 1314 | 149644 | 24.1 | 19601 | 6.0 |
| B | 692 | 105984 | 14.1 | 9939 | 2.4 |
| C | 364 | 99623 | 11.8 | 5308 | 1.5 |
| D | 1453 | 94221 | 10.3 | 21426 | 4.8 |
| E | 677 | 77381 | 8.5 | 9570 | 1.8 |
| H | 1849 | 98187 | 10.8 | 28425 | 4.8 |
| J | 712 | 107003 | 13.5 | 11383 | 2.3 |
| K | 0 | 1 | 0.0 | 1 | 0.0 |
| L | 159 | 179827 | 33.0 | 2555 | 1.0 |
| N | 47 | 57244 | 4.2 | 672 | 0.2 |
| P | 46 | 27357 | 1.9 | 638 | 0.1 |
| R | 36 | 6653 | 0.2 | 405 | 0.0 |
| U | 0 | 1 | 0.0 | 1 | 0.0 |

nb edges = topo before - 2 (except when topo before = 1)

tldr: always faster, really faster on L

### CODE REVIEW

**Comment by TeXitoi** [comment]
@l-vincent-l small improvement of the thermometer heuristic!

**Comment by TeXitoi** [comment]
For Transilien L on fr-idf, before 32s, after 2s (+ 5s of jormungandr).

File: source/time_tables/route_schedules.cpp [hunk 1]
```diff
@@ -177,37 +173,124 @@ std::vector<Edge> create_edges(std::vector<std::vector<datetime_stop_time>>& v)
...
+        size_t imax = edges.size();
+        while (imin < imax) {
+            const size_t imid = (imin + imax) / 2;
+
+            // modifying the graph to be in the wanted state
+            const size_t nb_elt_to_add = imid - done_until + 1;
+            if (e_descrs.size() < nb_elt_to_add) {
+                // add corresponding edges
+                while (nb_elt_to_add != e_descrs.size()) {
+                    const size_t idx = done_until + e_descrs.size();
```
Comment (antoine-de) [inline source/time_tables/route_schedules.cpp:259]: I find that writing it with a for loop is easier to understand (but maybe only for me :wink: ):

`for(const auto idx = done_until; idx < done_until + nb_elt_to_add; ++ idx)` no ?

File: source/time_tables/thermometer.cpp [hunk 2]
```diff
@@ -188,7 +189,9 @@ bool Thermometer::generate_topological_thermometer(const std::vector<vector_idx>
     return true;
 }
 
-void Thermometer::generate_thermometer(const std::vector<vector_idx> &stop_point_lists) {
+void Thermometer::generate_thermometer(const std::vector<vector_idx> &sps) {
+    std::vector<vector_idx> stop_point_lists = sps;
+    boost::range::sort(stop_point_lists, [](const vector_idx& a, const vector_idx&b) { return a.size() > b.size(); });
```
Comment (antoine-de) [inline source/time_tables/thermometer.cpp:196]: I think it's worth it to add a comment explaining why we do this.

File: source/time_tables/route_schedules.cpp [hunk 3]
```diff
@@ -177,37 +173,124 @@ std::vector<Edge> create_edges(std::vector<std::vector<datetime_stop_time>>& v)
...
-        } catch (boost::not_a_dag&) {
-            remove_edge(e_desc, g);
+    int nb_removed_edges = 0;
+    LOG4CPLUS_DEBUG(logger, "trying total ranked pair with nb_edges = " << edges.size());
+
+    // Straintforward implementation of ranked pair is:
+    //
+    //for (const auto& edge: edges) {
+    //    const auto e_desc = add_edge(edge.source, edge.target, g).first;
+    //    if (! is_dag(g)) { ++nb_removed_edges; remove_edge(e_desc, g); }
```
Comment (antoine-de) [inline source/time_tables/route_schedules.cpp]: the ++nb_removed_edges is not important for the comment no ?

Comment (TeXitoi) [inline source/time_tables/route_schedules.cpp]: That's just because it will do exactly the same thing.

File: source/time_tables/route_schedules.cpp (hunk 1 above)
Comment (TeXitoi) [inline source/time_tables/route_schedules.cpp:259]: I think the comment is addressed.

Comment (l-vincent-l) [inline source/time_tables/route_schedules.cpp]: I can read irony here.

Comment (TeXitoi) [inline source/time_tables/route_schedules.cpp]: Really not, we have clarified that together and make some modifications, but not here, thus github let the comment.
